src/numerical/spline.rs

```diff
@@ -36,71 +39,188 @@ use util::non_macro::*;
 /// // -2.2594x^3 + 14.2342x^2 - 28.5513x + 20.2094
 /// ```
 pub fn cubic_spline(node_x: Vector, node_y: Vector) -> Vec<Polynomial> {
-    //! Pre calculated variables
-    //! node_x: n+1
-    //! node_y: n+1
-    //! h     : n
-    //! b     : n
-    //! v     : n
-    //! u     : n
-    //! z     : n+1
-    let n = node_x.len() - 1;
-    assert_eq!(n, node_y.len() - 1);
-
-    // Pre-calculations
-    let mut h = vec![0f64; n];
-    let mut b = vec![0f64; n];
-    let mut v = vec![0f64; n];
-    let mut u = vec![0f64; n];
-    for i in 0..n {
-        if i == 0 {
-            h[i] = node_x[i + 1] - node_x[i];
-            b[i] = (node_y[i + 1] - node_y[i]) / h[i];
-        } else {
-            h[i] = node_x[i + 1] - node_x[i];
-            b[i] = (node_y[i + 1] - node_y[i]) / h[i];
-            v[i] = 2. * (h[i] + h[i - 1]);
-            u[i] = 6. * (b[i] - b[i - 1]);
+    let spline = CubicSpline::from_x_and_y(node_x, node_y);
+    spline.into()
+}
+
+/// Cubic Spline (Natural)
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct CubicSpline {
+    polynomials: Vec<(Range<f64>, Polynomial)>,
+}
+
+impl CubicSpline {
+    /// # Examples
+    /// ```
+    /// extern crate peroxide;
+    /// use peroxide::*;
+    ///
+    /// let x = c!(0.9, 1.3, 1.9, 2.1);
+    /// let y = c!(1.3, 1.5, 1.85, 2.1);
+    ///
+    /// let s = CubicSpline::from_x_and_y(x, y);
+    ///
+    /// for i in 0 .. 4 {
+    ///     println!("{}", s.eval(i as f64 / 2.0));
+    /// }
+    /// ```
+    pub fn from_x_and_y(node_x: Vector, node_y: Vector) -> Self {
```

Comment:
@Axect, I'm not so happy about the function name. What is your opinion here?

Reply:
@schrieveslaach Hmm.. I'm also uncomfortable too. Wouldn't `from_nodes`be better?

Reply:
That sounds good. I'll fix that later.